### PR TITLE
devicetree: document and test includes at child-binding level

### PR DESCRIPTION
**Pull request by gmarull**
It looks like it was possible to include other bindings at any level other than root. However, this feature was not tested nor documented.

Example:
```yaml
   # foo.yaml will be merged with content at this level
   include: foo.yaml
   child-binding:
     # bar.yaml will be merged with content at this level
     include: bar.yaml
```

### CODE REVIEW

File: dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml:104]: This one isn't in the allowlist that's being removed above, and therefore should not be here, right?

File: dts/bindings/pinctrl/pincfg-node-group.yaml
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/pincfg-node-group.yaml:8]: Don't you think we should keep this here, for out of tree users who are depending on it?

Not blocking on this one since I no longer maintain bindings, but I would strongly suggest it.

File: dts/bindings/pinctrl/pincfg-node.yaml
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/pincfg-node.yaml:19]: Again, you are breaking out of tree users by promoting these to top level.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Clever idea about the bindings language extension in the python-devicetree package. I am +1 for the idea, but please update the documentation in doc/build/dts/bindings.rst as well before we can merge this.

I found what looks like one mistake in the bindings changes.

Since the current changes to the bindings are breaking for out of tree users, I think this is pretty clearly a treewide change, so I'm going to apply that label. If you create a new bindings include file that's meant to be pulled in from the child-binding level instead, that would be backwards compatible and I'd be fine with removing the label in that case.

File: dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml
Comment (gmarull) [reply dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml:104]: Oops my mistake

**Comment by gmarull** [comment]
> Clever idea about the bindings language extension in the python-devicetree package. I am +1 for the idea, but please update the documentation in doc/build/dts/bindings.rst as well before we can merge this.
> 

Updated docs

> I found what looks like one mistake in the bindings changes.
>

Fixed
 
> Since the current changes to the bindings are breaking for out of tree users, I think this is pretty clearly a treewide change, so I'm going to apply that label. If you create a new bindings include file that's meant to be pulled in from the child-binding level instead, that would be backwards compatible and I'd be fine with removing the label in that case.

As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.

File: dts/bindings/pinctrl/pincfg-node-group.yaml
Comment (mbolivar-nordic) [reply dts/bindings/pinctrl/pincfg-node-group.yaml:8]: > I’d be surprised if there are pinctrl oot users, since it’s a de-facto requirement for any platform.

There are out of tree socs and even out-of-tree architectures. These are first class citizens in Zephyr. I understand you want to make changes quickly but I would not agree that the new treewide change policy is "unnecessary" and neither did the rest of the TSC.

**Comment by mbolivar-nordic** [comment]
> As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.

Fine, but this is still a treewide change and subject to the associated requirements.

**Review by mbolivar-nordic — DISMISSED** [review]
+1 from me on the DT infrastructure. Will add DNM just to be extra careful since it's treewide and hasn't gone through arch WG yet

**Comment by gmarull** [comment]
> > As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.
> 
> Fine, but this is still a treewide change and subject to the associated requirements.

why is something affecting a single driver class treewide?

**Comment by mbolivar-nordic** [comment]
> why is something affecting a single driver class treewide?

I think this is addressed by the definition of treewide here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

**Review by carlocaione — DISMISSED** [review]

**Comment by carlescufi** [comment]
> > why is something affecting a single driver class treewide?
> 
> I think this is addressed by the definition of treewide here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

I have to agree with @gmarull here. This is just too narrow in scope to fit that definition in my opinion.

**Review by galak — CHANGES_REQUESTED** [review]
Can you add more tests to verify that `property-allowlist` and `property-blocklist` work there.

**Comment by gmarull** [comment]
> Can you add more tests to verify that `property-allowlist` and `property-blocklist` work there.

done

**Comment by gmarull** [comment]
Added TSC label to discuss wether this applies for treewide or not.

**Comment by gmarull** [comment]
I realized we actually support this feature, but it was not documented or tested. Updated this PR with tests/docs only, and splitted pinctrl part here https://github.com/zephyrproject-rtos/zephyr/pull/52290. This is no longer a treewide change.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**
